diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 7c17093..267b34e 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -909,6 +909,12 @@ static int unionfs_setattr(struct dentry *dentry, struct iattr *ia)
 	}
 	lower_inode = unionfs_lower_inode(inode);
 
+        if (!lower_inode) {
+                // maybe this will fix the crash?
+                err = -EINVAL;
+                goto out;
+        }
+
 	/* check if user has permission to change lower inode */
 	err = inode_change_ok(lower_inode, ia);
 	if (err)
@@ -942,6 +948,11 @@ static int unionfs_setattr(struct dentry *dentry, struct iattr *ia)
 		/* get updated lower_dentry/inode after copyup */
 		lower_dentry = unionfs_lower_dentry(dentry);
 		lower_inode = unionfs_lower_inode(inode);
+
+                if (!lower_inode || !lower_dentry) {
+                    err = -ENOENT;
+                    goto out;
+                }
 	}
 
 	/*
